sycl/include/CL/sycl/half_type.hpp

```diff
@@ -26,6 +26,11 @@
 #else
 #define __SYCL_CONSTEXPR_ON_DEVICE
 #endif
+#if __cplusplus >= 201402L
```

Comment:
Why is there the sycl-include-gnu11.cpp test which checks it compiles with C++11? I thought we only support compiling as >=C++14.

Reply:
I think the RT library is C++14, but the headers are C++11.

Reply:
@bader , @romanovvlad - Can't we assume C++14 or newer in SYCL header files?

Reply:
I don't have a simple "yes/no" answer, but I'm aware of a couple of factors restricting usage of C++ features:
- SYCL specification: SYCL-1.2.1 relies on C++11 and SYCL-2020 relies on C++17 features. It seems to me that DPC++ is not going to support "strict SYCL-1.2.1" mode where C++14 (or later) features are not supported.
- Runtime library ABI compatibility (see the note in our [contribution guide](https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#development)). It's hard to say if using C++14 features in SYCL headers breaks runtime ABI, but it possible and should be prevented by regression tests.

Reply:
We have C++ version documented at:
https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#c-standard
but it doesn't answer whether C++11 mode will be functional or broken.

I think it would be good to know and catch the case when we really want to break C++11 compatibility and have some discussion on that. Maybe that's what sycl-include-gnu11.cpp test is doing?

Reply:
The test only checks that the header can be included. But not that any template in the header can be instantiated in C++11 mode. The current version of this PR passes all tests . But trying to instantiate known_identity for half in C++11 will give a compiler error (not tested).

It shouldn't have any ABI issues to use any C++14 features in the header. The problematic release for ABI issues is C++11 (e.g. string). There weren't any such things in C++14.

Reply:
C++14 is required now. The doc changes have been merged 10 days ago: https://github.com/intel/llvm/pull/3053